# Split Learning for Graph Neural Network

>The following codes are demos only. It's **NOT for production** due to system security concerns, please **DO NOT** use it directly in production.

## Setup

Create two participant alice and bob.

In [1]:
import secretflow as sf

# Check the version of your SecretFlow
print('The version of SecretFlow: {}'.format(sf.__version__))

# In case you got a running secetflow runtime already.
sf.shutdown()

sf.init(parties=['alice', 'bob'], address='local')

alice, bob = sf.PYU('alice'), sf.PYU('bob')

The version of SecretFlow: 1.7.0b0


/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/subprocess.py:1796: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = _posixsubprocess.fork_exec(
2024-08-04 16:10:02,927	INFO worker.py:1724 -- Started a local Ray instance.


(pid=1985337) 2024-08-04 16:13:34.404821: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
(pid=1985337) 2024-08-04 16:13:34.429374: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
(pid=1985337) 2024-08-04 16:13:34.437930: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
(pid=1985337) 2024-08-04 16:13:35.482220: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


(PYUSLTFModel pid=1985337) Model: "functional"
(PYUSLTFModel pid=1985337) ┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
(PYUSLTFModel pid=1985337) ┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
(PYUSLTFModel pid=1985337) ┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
(PYUSLTFModel pid=1985337) │ input_layer         │ (None, 716)       │          0 │ -                 │
(PYUSLTFModel pid=1985337) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1985337) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid=1985337) │ input_layer_1       │ (None, 2708)      │          0 │ -                 │
(PYUSLTFModel pid=1985337) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1985337) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid

(PYUSLTFModel pid=1985337) Exception raised in creation task: The actor died because of an error raised in its creation task, ray::PYUSLTFModel.__init__() (pid=1985337, ip=10.0.0.3, actor_id=c66dbf62be30d2035ddb4ad801000000, repr=<secretflow.ml.nn.sl.backend.tensorflow.sl_base.PYUSLTFModel object at 0x7fc3fd4879d0>)
(PYUSLTFModel pid=1985337)   File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/secretflow/ml/nn/sl/backend/tensorflow/sl_base.py", line 88, in __init__
(PYUSLTFModel pid=1985337)     self.model_fuse = builder_fuse() if builder_fuse is not None else None
(PYUSLTFModel pid=1985337)   File "/tmp/ipykernel_1983960/2114966435.py", line 7, in fuse_model
(PYUSLTFModel pid=1985337)   File "/tmp/ipykernel_1983960/2114966435.py", line 7, in <listcomp>
(PYUSLTFModel pid=1985337)   File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/keras/src/layers/core/input_layer.py", line 148, in Input
(PYUSLTFModel pid=1985337)     layer = InputLayer(
(PYUSLTF

(PYUSLTFModel pid=1985488) Model: "functional"
(PYUSLTFModel pid=1985488) ┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
(PYUSLTFModel pid=1985488) ┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
(PYUSLTFModel pid=1985488) ┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
(PYUSLTFModel pid=1985488) │ input_layer         │ (None, 716)       │          0 │ -                 │
(PYUSLTFModel pid=1985488) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1985488) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid=1985488) │ input_layer_1       │ (None, 2708)      │          0 │ -                 │
(PYUSLTFModel pid=1985488) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1985488) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid

(PYUSLTFModel pid=1985488) Exception raised in creation task: The actor died because of an error raised in its creation task, ray::PYUSLTFModel.__init__() (pid=1985488, ip=10.0.0.3, actor_id=65056c5b3736e4570ff8bc8901000000, repr=<secretflow.ml.nn.sl.backend.tensorflow.sl_base.PYUSLTFModel object at 0x7f29c4a871f0>)
(PYUSLTFModel pid=1985488)   File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/secretflow/ml/nn/sl/backend/tensorflow/sl_base.py", line 88, in __init__
(PYUSLTFModel pid=1985488)     self.model_fuse = builder_fuse() if builder_fuse is not None else None
(PYUSLTFModel pid=1985488)   File "/tmp/ipykernel_1983960/2114966435.py", line 7, in fuse_model
(PYUSLTFModel pid=1985488)   File "/tmp/ipykernel_1983960/2114966435.py", line 7, in <listcomp>
(PYUSLTFModel pid=1985488)   File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/keras/src/layers/core/input_layer.py", line 148, in Input
(PYUSLTFModel pid=1985488)     layer = InputLayer(
(PYUSLTF

(PYUSLTFModel pid=1986097) Model: "functional"
(PYUSLTFModel pid=1986097) ┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
(PYUSLTFModel pid=1986097) ┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
(PYUSLTFModel pid=1986097) ┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
(PYUSLTFModel pid=1986097) │ input_layer         │ (None, 717)       │          0 │ -                 │
(PYUSLTFModel pid=1986097) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1986097) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid=1986097) │ input_layer_1       │ (None, 10)        │          0 │ -                 │
(PYUSLTFModel pid=1986097) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1986097) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid

(pid=1986397) 2024-08-04 16:17:26.591199: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
(pid=1986397) 2024-08-04 16:17:26.614401: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
(pid=1986397) 2024-08-04 16:17:26.623978: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
(pid=1986397) 2024-08-04 16:17:27.683781: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


(PYUSLTFModel pid=1986429) Model: "functional"
(PYUSLTFModel pid=1986429) ┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
(PYUSLTFModel pid=1986429) ┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
(PYUSLTFModel pid=1986429) ┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
(PYUSLTFModel pid=1986429) │ input_layer         │ (None, 717)       │          0 │ -                 │
(PYUSLTFModel pid=1986429) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1986429) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid=1986429) │ input_layer_1       │ (None, 2708)      │          0 │ -                 │
(PYUSLTFModel pid=1986429) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1986429) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid

## Prepare the Dataset

### The cora dataset
The [cora](https://linqs-data.soe.ucsc.edu/public/lbc/cora.tgz) dataset has two tap-separated files: `cora.cites` and `cora.content`.

- The `cora.cites` includes the citation records with two columns: cited_paper_id (target) and citing_paper_id (source).
- The `cora.content` includes the paper content records with 1,435 columns: paper_id, subject, and 1,433 binary features.

Let us use the partitioned cora dataset, which is already a built-in dataset of SecretFlow.

- The train set includes 140 cited_paper_ids.
- The test set includes 1000 cited_paper_ids.
- The valid set includes 500 cited_paper_ids.

### Split the dataset

Let us split the dataset for split learning.

- Alice holds the 1~716 features, and bob holds the left.
- Alice holds all label.
- Alice and bob hold all edges both.

In [2]:
import networkx as nx
import numpy as np
import os
import pickle
import scipy
import zipfile
from pathlib import Path
from secretflow.utils.simulation.datasets import dataset
from secretflow.data.ndarray import load


def load_cora():
    dataset_zip = dataset('cora')
    extract_path = str(Path(dataset_zip).parent)
    with zipfile.ZipFile(dataset_zip, 'r') as zip_f:
        zip_f.extractall(extract_path)

    file_names = [
        os.path.join(extract_path, f'ind.cora.{name}')
        for name in ['y', 'tx', 'ty', 'allx', 'ally', 'graph']
    ]

    objects = []
    for name in file_names:
        with open(name, 'rb') as f:
            objects.append(pickle.load(f, encoding='latin1'))

    y, tx, ty, allx, ally, graph = tuple(objects)

    with open(os.path.join(extract_path, f"ind.cora.test.index"), 'r') as f:
        test_idx_reorder = f.readlines()
    test_idx_reorder = list(map(lambda s: int(s.strip()), test_idx_reorder))
    test_idx_range = np.sort(test_idx_reorder)

    nodes = scipy.sparse.vstack((allx, tx)).tolil()
    nodes[test_idx_reorder, :] = nodes[test_idx_range, :]
    edge = nx.adjacency_matrix(nx.from_dict_of_lists(graph))
    edge = edge.toarray() + np.eye(edge.shape[1])

    labels = np.vstack((ally, ty))
    labels[test_idx_reorder, :] = labels[test_idx_range, :]

    idx_test = test_idx_range.tolist()
    idx_train = range(len(y))
    idx_val = range(len(y), len(y) + 500)

    def sample_mask(idx, length):
        mask = np.zeros(length)
        mask[idx] = 1
        return np.array(mask, dtype=bool)

    idx_train = sample_mask(idx_train, labels.shape[0])
    idx_val = sample_mask(idx_val, labels.shape[0])
    idx_test = sample_mask(idx_test, labels.shape[0])

    y_train = np.zeros(labels.shape)
    y_val = np.zeros(labels.shape)
    y_test = np.zeros(labels.shape)
    y_train[idx_train, :] = labels[idx_train, :]
    y_val[idx_val, :] = labels[idx_val, :]
    y_test[idx_test, :] = labels[idx_test, :]

    nodes = nodes.toarray()
    features_split_pos = round(nodes.shape[1] / 2)
    nodes_alice, nodes_bob = (
        nodes[:, :features_split_pos],
        nodes[:, features_split_pos:],
    )
    temp_dir = "/home/beng003/python_project/sf-test/data"
    saved_files = [
        os.path.join(temp_dir, name)
        for name in [
            'edge.npy',
            'x_alice.npy',
            'x_bob.npy',
            'y_train.npy',
            'y_val.npy',
            'y_test.npy',
            'idx_train.npy',
            'idx_val.npy',
            'idx_test.npy',
        ]
    ]
    np.save(saved_files[0], edge)
    np.save(saved_files[1], nodes_alice)
    np.save(saved_files[2], nodes_bob)
    np.save(saved_files[3], y_train)
    np.save(saved_files[4], y_val)
    np.save(saved_files[5], y_test)
    np.save(saved_files[6], idx_train)
    np.save(saved_files[7], idx_val)
    np.save(saved_files[8], idx_test)
    return saved_files

saved_files = load_cora()

edge = load({alice: saved_files[0], bob: saved_files[0]})
features = load({alice: saved_files[1], bob: saved_files[2]})
Y_train = load({alice: saved_files[3]})
Y_val = load({alice: saved_files[4]})
Y_test = load({alice: saved_files[5]})
idx_train = load({alice: saved_files[6]})
idx_val = load({alice: saved_files[7]})
idx_test = load({alice: saved_files[8]})

/tmp/ipykernel_1983960/977647887.py:26: DeprecationWarning: Please import `csr_matrix` from the `scipy.sparse` namespace; the `scipy.sparse.csr` namespace is deprecated and will be removed in SciPy 2.0.0.
  objects.append(pickle.load(f, encoding='latin1'))


By the way, since cora is a built-in dataset of SecretFlow, you can just run the following snippet to replace the codes above.

In [17]:
from secretflow.utils.simulation.datasets import load_cora

(edge, features, Y_train, Y_val, Y_test, idx_train, idx_val, idx_test) = load_cora(
    [alice, bob]
)
# edge = edge[0:10]
# features = features[0:10]
# Y_train = Y_train[0:10]
# Y_val = Y_val[0:10]
# Y_test = Y_test[0:10]
# idx_train = idx_train[0:10]
# idx_val = idx_val[0:10]
# idx_test = idx_test[0:10]


## Build a Graph Neural Network Model


### Implement a graph convolution layer

In [7]:
import tensorflow as tf
from tensorflow.keras import activations
from tensorflow.keras import backend as K
from tensorflow.keras import constraints, initializers, regularizers
from tensorflow.keras.layers import Dropout, Layer, LeakyReLU


class GraphAttention(Layer):
    def __init__(
        self,
        F_,
        attn_heads=1,
        attn_heads_reduction='average',  # {'concat', 'average'}
        dropout_rate=0.5,
        activation='relu',
        use_bias=True,
        kernel_initializer='glorot_uniform',
        bias_initializer='zeros',
        attn_kernel_initializer='glorot_uniform',
        kernel_regularizer=None,
        bias_regularizer=None,
        attn_kernel_regularizer=None,
        activity_regularizer=None,
        kernel_constraint=None,
        bias_constraint=None,
        attn_kernel_constraint=None,
        **kwargs,
    ):
        if attn_heads_reduction not in {'concat', 'average'}:
            raise ValueError('Possbile reduction methods: concat, average')

        self.F_ = F_  # Number of output features (F' in the paper)
        self.attn_heads = attn_heads  # Number of attention heads (K in the paper)
        self.attn_heads_reduction = attn_heads_reduction
        self.dropout_rate = dropout_rate  # Internal dropout rate
        self.activation = activations.get(activation)
        self.use_bias = use_bias

        self.kernel_initializer = initializers.get(kernel_initializer)
        self.bias_initializer = initializers.get(bias_initializer)
        self.attn_kernel_initializer = initializers.get(attn_kernel_initializer)

        self.kernel_regularizer = regularizers.get(kernel_regularizer)
        self.bias_regularizer = regularizers.get(bias_regularizer)
        self.attn_kernel_regularizer = regularizers.get(attn_kernel_regularizer)
        self.activity_regularizer = regularizers.get(activity_regularizer)

        self.kernel_constraint = constraints.get(kernel_constraint)
        self.bias_constraint = constraints.get(bias_constraint)
        self.attn_kernel_constraint = constraints.get(attn_kernel_constraint)
        self.supports_masking = False

        # Populated by build()
        self.kernels = []  # Layer kernels for attention heads
        self.biases = []  # Layer biases for attention heads
        self.attn_kernels = []  # Attention kernels for attention heads

        if attn_heads_reduction == 'concat':
            # Output will have shape (..., K * F')
            self.output_dim = self.F_ * self.attn_heads
        else:
            # Output will have shape (..., F')
            self.output_dim = self.F_

        super(GraphAttention, self).__init__(**kwargs)

    def build(self, input_shape):
        assert len(input_shape) >= 2
        F = input_shape[0][-1]

        # Initialize weights for each attention head
        for head in range(self.attn_heads):
            # Layer kernel
            kernel = self.add_weight(
                shape=(F, self.F_),
                initializer=self.kernel_initializer,
                regularizer=self.kernel_regularizer,
                constraint=self.kernel_constraint,
                name='kernel_{}'.format(head),
            )
            self.kernels.append(kernel)

            # # Layer bias
            if self.use_bias:
                bias = self.add_weight(
                    shape=(self.F_,),
                    initializer=self.bias_initializer,
                    regularizer=self.bias_regularizer,
                    constraint=self.bias_constraint,
                    name='bias_{}'.format(head),
                )
                self.biases.append(bias)

            # Attention kernels
            attn_kernel_self = self.add_weight(
                shape=(self.F_, 1),
                initializer=self.attn_kernel_initializer,
                regularizer=self.attn_kernel_regularizer,
                constraint=self.attn_kernel_constraint,
                name='attn_kernel_self_{}'.format(head),
            )
            attn_kernel_neighs = self.add_weight(
                shape=(self.F_, 1),
                initializer=self.attn_kernel_initializer,
                regularizer=self.attn_kernel_regularizer,
                constraint=self.attn_kernel_constraint,
                name='attn_kernel_neigh_{}'.format(head),
            )
            self.attn_kernels.append([attn_kernel_self, attn_kernel_neighs])
        self.built = True

    def call(self, inputs):
        X = inputs[0]  # Node features (N x F)
        A = inputs[1]  # Adjacency matrix (N x N)

        outputs = []
        for head in range(self.attn_heads):
            kernel = self.kernels[head]  # W in the paper (F x F')
            attention_kernel = self.attn_kernels[
                head
            ]  # Attention kernel a in the paper (2F' x 1)

            # Compute inputs to attention network
            features = K.dot(X, kernel)  # (N x F')

            # Compute feature combinations
            # Note: [[a_1], [a_2]]^T [[Wh_i], [Wh_2]] = [a_1]^T [Wh_i] + [a_2]^T [Wh_j]
            attn_for_self = K.dot(
                features, attention_kernel[0]
            )  # (N x 1), [a_1]^T [Wh_i]
            attn_for_neighs = K.dot(
                features, attention_kernel[1]
            )  # (N x 1), [a_2]^T [Wh_j]

            # Attention head a(Wh_i, Wh_j) = a^T [[Wh_i], [Wh_j]]
            dense = attn_for_self + K.transpose(
                attn_for_neighs
            )  # (N x N) via broadcasting

            # Add nonlinearty
            dense = LeakyReLU(alpha=0.2)(dense)

            # Mask values before activation (Vaswani et al., 2017)
            mask = -10e9 * (1.0 - A)
            dense += mask

            # Apply softmax to get attention coefficients
            dense = K.softmax(dense)  # (N x N)

            # Apply dropout to features and attention coefficients
            dropout_attn = Dropout(self.dropout_rate)(dense)  # (N x N)
            dropout_feat = Dropout(self.dropout_rate)(features)  # (N x F')

            # Linear combination with neighbors' features
            node_features = K.dot(dropout_attn, dropout_feat)  # (N x F')

            if self.use_bias:
                node_features = K.bias_add(node_features, self.biases[head])

            # Add output of attention head to final output
            outputs.append(node_features)

        # Aggregate the heads' output according to the reduction method
        if self.attn_heads_reduction == 'concat':
            output = K.concatenate(outputs)  # (N x KF')
        else:
            output = K.mean(K.stack(outputs), axis=0)  # N x F')

        output = self.activation(output)
        return output

    def compute_output_shape(self, input_shape):
        output_shape = input_shape[0][0], self.output_dim
        return output_shape

    def get_config(self):
        config = super().get_config().copy()
        config.update(
            {
                'attn_heads': self.attn_heads,
                'attn_heads_reduction': self.attn_heads_reduction,
                'F_': self.F_,
            }
        )
        return config

2024-08-04 16:12:57.164535: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-08-04 16:12:57.185096: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-08-04 16:12:57.191085: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-08-04 16:12:58.274560: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


### Implement a fuse net
The fuse model is used in the party with the label. It works as follows:
1. Use the concated node embeddings to generat the final node embeddings.
2. Feed the node embeddings in a Softmax layer to predict the node class.

In [8]:
class ServerNet(tf.keras.layers.Layer):
    def __init__(
        self,
        in_channel: int,
        hidden_size: int,
        num_layer: int,
        num_class: int,
        dropout: float,
        **kwargs,
    ):
        super(ServerNet, self).__init__()
        self.num_class = num_class
        self.num_layer = num_layer
        self.hidden_size = hidden_size
        self.in_channel = in_channel
        self.dropout = dropout
        self.layers = []
        super(ServerNet, self).__init__(**kwargs)

    def build(self, input_shape):
        self.layers.append(
            tf.keras.layers.Dense(self.hidden_size, input_shape=(self.in_channel,))
        )
        for i in range(self.num_layer - 2):
            self.layers.append(
                tf.keras.layers.Dense(self.hidden_size, input_shape=(self.hidden_size,))
            )
        self.layers.append(
            tf.keras.layers.Dense(self.num_class, input_shape=(self.hidden_size,))
        )

        super(ServerNet, self).build(input_shape)

    def call(self, inputs):
        x = inputs
        x = Dropout(self.dropout)(x)
        for i in range(self.num_layer):
            x = Dropout(self.dropout)(x)
            x = self.layers[i](x)

        return K.softmax(x)

    def compute_output_shape(self, input_shape):
        output_shape = self.hidden_size, self.output_dim
        return output_shape

    def get_config(self):
        config = super().get_config().copy()
        config.update(
            {
                'in_channel': self.in_channel,
                'hidden_size': self.hidden_size,
                'num_layer': self.num_layer,
                'num_class': self.num_class,
                'dropout': self.dropout,
            }
        )
        return config

### Build the base model
The base model is used in each party to generate node embeddings. It applys one graph convolutional layer to produce node embeddings.

The node embeddings of all parties are then transfered to the party with labels for further processing.

In [9]:
from tensorflow.keras.models import Model


def create_base_model(
    input_shape, n_hidden, l2_reg, num_heads, dropout_rate, learning_rate
):
    def base_model():
        feature_input = tf.keras.Input(shape=(input_shape[1],))
        graph_input = tf.keras.Input(shape=(input_shape[0],))
        regular = tf.keras.regularizers.l2(l2_reg)
        outputs = GraphAttention(
            F_=n_hidden,
            attn_heads=num_heads,
            attn_heads_reduction='average',  # {'concat', 'average'}
            dropout_rate=dropout_rate,
            activation='relu',
            use_bias=True,
            kernel_initializer='glorot_uniform',
            bias_initializer='zeros',
            attn_kernel_initializer='glorot_uniform',
            kernel_regularizer=regular,
            bias_regularizer=None,
            attn_kernel_regularizer=None,
            activity_regularizer=None,
            kernel_constraint=None,
            bias_constraint=None,
            attn_kernel_constraint=None,
        )([feature_input, graph_input])
        # outputs = tf.keras.layers.Flatten()(outputs)
        model = Model(inputs=[feature_input, graph_input], outputs=outputs)
        model._name = "embed_model"
        # Compile model
        model.summary()
        metrics = ['acc']
        optimizer = tf.keras.optimizers.get(
            {
                'class_name': 'adam',
                'config': {'learning_rate': learning_rate},
            }
        )
        model.compile(
            loss='categorical_crossentropy',
            weighted_metrics=metrics,
            optimizer=optimizer,
        )
        return model

    return base_model

### Build the fuse model
The fuse model concat the node embeddings from all parties as input. It works only in the party with the label.

In [10]:
from tensorflow import keras
from tensorflow.keras import layers


def create_fuse_model(hidden_units, hidden_size, n_classes, layer_num, learning_rate):
    def fuse_model():
        inputs = [keras.Input(shape=size) for size in hidden_units]
        x = layers.concatenate(inputs)
        input_shape = x.shape[-1]
        y_pred = ServerNet(
            in_channel=input_shape,
            hidden_size=hidden_size,
            num_layer=layer_num,
            num_class=n_classes,
            dropout=0.0,
        )(x)
        # Create the model.
        model = keras.Model(inputs=inputs, outputs=y_pred, name="fuse_model")
        model.summary()
        metrics = ['acc']
        optimizer = tf.keras.optimizers.get(
            {
                'class_name': 'adam',
                'config': {'learning_rate': learning_rate},
            }
        )
        model.compile(
            loss='categorical_crossentropy',
            weighted_metrics=metrics,
            optimizer=optimizer,
        )
        return model

    return fuse_model

## Train GNN model with split learning

Let us build a split learning model for training. 

Alice who has the label holds a base model and a fuse model, while bob holds a base model only.

The whole model structure is as follow

<img alt="split_learning_gnn_model.png" src="resources/split_gnn.svg" width="400">

In [19]:
from secretflow.ml.nn import SLModel

hidden_size = 256
n_classes = 7
attn_heads = 2
layer_num = 3
learning_rate = 1e-3
dropout_rate = 0.0
l2_reg = 0.1
num_heads = 4
epochs = 10
optimizer = 'adam'

partition_shapes = features.partition_shape()

input_shape_alice = partition_shapes[alice]
input_shape_bob = partition_shapes[bob]

sl_model = SLModel(
    base_model_dict={
        alice: create_base_model(
            input_shape_alice,
            hidden_size,
            l2_reg,
            num_heads,
            dropout_rate,
            learning_rate,
        ),
        bob: create_base_model(
            input_shape_bob,
            hidden_size,
            l2_reg,
            num_heads,
            dropout_rate,
            learning_rate,
        ),
    },
    device_y=alice,
    model_fuse=create_fuse_model(
        [hidden_size, hidden_size], hidden_size, n_classes, layer_num, learning_rate
    ),
)

INFO:root:Create proxy actor <class 'secretflow.ml.nn.sl.backend.tensorflow.sl_base.PYUSLTFModel'> with party alice.


INFO:root:Create proxy actor <class 'secretflow.ml.nn.sl.backend.tensorflow.sl_base.PYUSLTFModel'> with party bob.


Fit the model.

In [20]:
sl_model.fit(
    x=[features, edge],
    y=Y_train,
    epochs=epochs,
    batch_size=input_shape_alice[0],
    sample_weight=idx_train,
    validation_data=([features, edge], Y_val, idx_val),
)

INFO:root:SL Train Params: {'self': <secretflow.ml.nn.sl.sl_model.SLModel object at 0x7f9f584c14e0>, 'x': [FedNdarray(partitions={PYURuntime(alice): <secretflow.device.device.pyu.PYUObject object at 0x7f9f584e24a0>, PYURuntime(bob): <secretflow.device.device.pyu.PYUObject object at 0x7f9f584e0fa0>}, partition_way=<PartitionWay.VERTICAL: 'vertical'>), FedNdarray(partitions={PYURuntime(alice): <secretflow.device.device.pyu.PYUObject object at 0x7f9f584c3f40>, PYURuntime(bob): <secretflow.device.device.pyu.PYUObject object at 0x7f9f584e3a30>}, partition_way=<PartitionWay.HORIZONTAL: 'horizontal'>)], 'y': FedNdarray(partitions={PYURuntime(alice): <secretflow.device.device.pyu.PYUObject object at 0x7f9f584e0ca0>}, partition_way=<PartitionWay.HORIZONTAL: 'horizontal'>), 'batch_size': 2708, 'epochs': 1, 'verbose': 1, 'callbacks': None, 'validation_data': ([FedNdarray(partitions={PYURuntime(alice): <secretflow.device.device.pyu.PYUObject object at 0x7f9f584e24a0>, PYURuntime(bob): <secretflow.

(PYUSLTFModel pid=1986397) Model: "functional"
(PYUSLTFModel pid=1986397) ┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
(PYUSLTFModel pid=1986397) ┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
(PYUSLTFModel pid=1986397) ┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
(PYUSLTFModel pid=1986397) │ input_layer         │ (None, 716)       │          0 │ -                 │
(PYUSLTFModel pid=1986397) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1986397) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid=1986397) │ input_layer_1       │ (None, 2708)      │          0 │ -                 │
(PYUSLTFModel pid=1986397) │ (InputLayer)        │                   │            │                   │
(PYUSLTFModel pid=1986397) ├─────────────────────┼───────────────────┼────────────┼───────────────────┤
(PYUSLTFModel pid

(PYUSLTFModel pid=1986397) Exception raised in creation task: The actor died because of an error raised in its creation task, ray::PYUSLTFModel.__init__() (pid=1986397, ip=10.0.0.3, actor_id=5539767725861cdc810ef78a01000000, repr=<secretflow.ml.nn.sl.backend.tensorflow.sl_base.PYUSLTFModel object at 0x7f9affe8edd0>)
(PYUSLTFModel pid=1986397)   File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/secretflow/ml/nn/sl/backend/tensorflow/sl_base.py", line 88, in __init__
(PYUSLTFModel pid=1986397)     self.model_fuse = builder_fuse() if builder_fuse is not None else None
(PYUSLTFModel pid=1986397)   File "/tmp/ipykernel_1983960/2114966435.py", line 7, in fuse_model
(PYUSLTFModel pid=1986397)   File "/tmp/ipykernel_1983960/2114966435.py", line 7, in <listcomp>
(PYUSLTFModel pid=1986397)   File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/keras/src/layers/core/input_layer.py", line 148, in Input
(PYUSLTFModel pid=1986397)     layer = InputLayer(
(PYUSLTF

RayActorError: The actor died because of an error raised in its creation task, ray::PYUSLTFModel.__init__() (pid=1986397, ip=10.0.0.3, actor_id=5539767725861cdc810ef78a01000000, repr=<secretflow.ml.nn.sl.backend.tensorflow.sl_base.PYUSLTFModel object at 0x7f9affe8edd0>)
  File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/secretflow/ml/nn/sl/backend/tensorflow/sl_base.py", line 88, in __init__
    self.model_fuse = builder_fuse() if builder_fuse is not None else None
  File "/tmp/ipykernel_1983960/2114966435.py", line 7, in fuse_model
  File "/tmp/ipykernel_1983960/2114966435.py", line 7, in <listcomp>
  File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/keras/src/layers/core/input_layer.py", line 148, in Input
    layer = InputLayer(
  File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/keras/src/layers/core/input_layer.py", line 47, in __init__
    shape = backend.standardize_shape(shape)
  File "/home/beng003/anaconda/envs/sf_1.7/lib/python3.10/site-packages/keras/src/backend/common/variables.py", line 530, in standardize_shape
    raise ValueError(f"Cannot convert '{shape}' to a shape.")
ValueError: Cannot convert '256' to a shape.

Examine the GNN model predictions

In [10]:
sl_model.evaluate(
    x=[features, edge],
    y=Y_test,
    batch_size=input_shape_alice[0],
    sample_weight=idx_test,
)

Evaluate Processing:: 100%|██████████| 1/1 [00:00<00:00,  2.26it/s, loss:0.4411766827106476 acc:0.7720000147819519]


{'loss': 0.44117668, 'acc': 0.772}

## Conclusion

In this tutorial, we demonstrate how to train graph neural network with split learning. This is a very basic implementation and there are some works we will explore in the future:

- SGD on large graphs: in the example above, in each training step, we have to do `prepare`, `aggregate` and `update` on whole graph, which is extremely computation intensive. We should perform stochastic minibatch training to reduce computation and memory comsumption.
- Partially aligned graphs: in the example above, parties must have same nodes set which may not be satisfied in real cases. We want to explore the case where all parties have common subset nodes.